File: src/foam/lib/parse/ErrorReportingPStream.java
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package foam.lib.parse;

import org.apache.commons.lang3.StringUtils;

import java.util.HashSet;
import java.util.Set;

public class ErrorReportingPStream
    extends ProxyPStream
{

  protected Parser errParser = null;
  protected ParserContext errContext = null;
  protected ErrorReportingNodePStream errStream = null;

  protected ErrorReportingNodePStream tail_ = null;
  protected Set<String> validCharacters = new HashSet<>();

  public ErrorReportingPStream(PStream delegate) {
    setDelegate(delegate);
  }

  @Override
  public PStream tail() {
    // tail becomes new node with increased position
    if ( tail_ == null ) tail_ = new ErrorReportingNodePStream(this, super.tail(), 1);
    return tail_;
  }

  @Override
  public PStream setValue(Object value) {
    // create a new node
    return new ErrorReportingNodePStream(this, super.setValue(value), 0);
  }

  @Override
  public PStream apply(Parser parser, ParserContext x) {
    PStream result = parser.parse(this, x);
    if ( result == null ) {
      // if result is null then self report
      this.report(new ErrorReportingNodePStream(this, getDelegate(), 0), parser, x);
    }
    return result;
  }

  public void report(ErrorReportingNodePStream ernps, Parser parser, ParserContext x) {
    // get the report with the furthest position
    if ( errStream == null || errStream.pos_ <= ernps.pos_ ) {
      errStream = ernps;
      errParser = parser;
      errContext = x;
    }
  }

  public void reportValidCharacter(Character ch) {
    validCharacters.add("'" + getPrintableCharacter(ch) + "'");
  }

  public String getMessage() {
    // check if err is valid and print the char, if not print EOF
    String invalid = ( errStream.valid() ) ? String.valueOf(errStream.head()) : "EOF";

    // get a list of valid characters
    TrapPStream trap = new TrapPStream(this);
    for ( int i = 0; i <= 255; i++ ) {
      trap.setHead((char) i);
      trap.apply(errParser, errContext);
    }

    return "Invalid character '" + invalid +
        "' found at " + errStream.pos_ + "\n" +
        "Valid characters include: " +
        StringUtils.join(validCharacters, ",");
  }

  public String getPrintableCharacter(Character ch) {
    switch ( (int) ch ) {
      case 0x09:
        return "\\t";
      case 0x0A:
        return "\\n";
      case 0x0D:
        return "\\r";
      default:
        return ch.toString();
    }
  }
}

File: src/foam/parse/ErrorReportingPStream.js
/**
 * @license
 * Copyright 2019 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.CLASS({
  package: 'foam.parse',
  name: 'ErrorReportingPStream',
  properties: [
    {
      class: 'Simple',
      name: 'delegate'
    },
    {
      name: 'intro'
    },
    {
      name: 'pos',
      value: 0
    },
    {
      name: 'limit',
      value: 100
    },
    {
      name: 'head',
      getter: function() { return this.delegate.head; }
    },
    {
      name: 'tail',
      getter: function() {
        if ( ! this.instance_.tail ) {
          var ps = this.cls_.create();
          ps.delegate = this.delegate.tail;
          ps.report = this.report;
          ps.pos = this.pos + 1;
          ps.limit = this.limit;
          ps.intro = ps.pos > ps.limit ?
            this.intro.tail : this.intro;
          this.instance_.tail = ps;
        }
        return this.instance_.tail;
      },
      setter: function(tail) { this.instance_.tail = tail; }
    },
    {
      name: 'valid',
      getter: function() { return this.delegate.valid; }
    },
    {
      name: 'value',
      setter: function(value) { this.delegate.value = value; },
      getter: function() { return this.delegate.value; }
    },
    {
      class: 'Simple',
      name: 'report'
    }
  ],

  methods: [
    function init() {
      this.intro = this;
    },

    function setValue(value) {
      if ( value === undefined ) value = null;
      var ps = this.cls_.create();
      ps.pos      = this.pos;
      ps.limit    = this.limit;
      ps.intro    = this.intro;
      ps.report   = this.report;
      ps.delegate = this.delegate.setValue(value);
      ps.tail     = this.tail;
      return ps;
    },

    function substring(end) {
      return this.delegate.substring(end.delegate);
    },

    function compareTo(o) {
      return this.delegate.compareTo(o.delegate);
    },

    function getIntroString() {
      return this.intro.substring(this);
    },

    function apply(p, obj) {
      var res = p.parse(this, obj);
      if ( ! res ) this.report(this, p, obj);
      return res;
    }
  ]
});
